src/wallet/wallet2.cpp

```diff
@@ -4983,129 +4948,126 @@ void wallet2::restore(const std::string& wallet_, const epee::wipeable_string& p
-    cryptonote::generate_multisig_N_N(get_account().get_keys(), spend_keys, multisig_keys, spend_skey, spend_pkey);
+    // 2. duplicate signers not allowed
+    CHECK_AND_ASSERT_THROW_MES(std::find(signers.begin(), signers.end(), expanded_msgs.back().get_signing_pubkey()) == signers.end(),
+      "Duplicate signers not allowed when converting a wallet to multisig.");
 
-    // Our signer key is b * G, where b is secret spend key.
-    multisig_signers = spend_keys;
-    multisig_signers.push_back(get_multisig_signer_public_key(get_account().get_keys().m_spend_secret_key));
+    // add signer (skip self for now)
+    if (expanded_msgs.back().get_signing_pubkey() != own_pubkey)
+      signers.push_back(expanded_msgs.back().get_signing_pubkey());
   }
-  else
-  {
-    // We just got public spend keys of all participants and deriving multisig keys (set of Mi = b * Bi).
-    // note that derivations are public keys as DH exchange suppose it to be
-    auto derivations = cryptonote::generate_multisig_derivations(get_account().get_keys(), spend_keys);
 
-    spend_pkey = rct::identity();
-    multisig_signers = std::vector<crypto::public_key>(spend_keys.size() + 1, crypto::null_pkey);
+  // add self to signers
+  signers.push_back(own_pubkey);
 
-    if (threshold == spend_keys.size())
-    {
-      // N - 1 / N case
+  // sort signers
+  std::sort(signers.begin(), signers.end(),
+      [](const crypto::public_key &a, const crypto::public_key &b) -> bool
+      {
+        return memcmp(&a, &b, sizeof(a)) < 0;
+      }
+    );
 
-      // We need an extra step, so we package all the composite public keys
-      // we know about, and make a signed string out of them
-      MINFO("Creating spend key...");
+  // additional checks
+  CHECK_AND_ASSERT_THROW_MES(signers.size() > 1,
+    "Tried to make multisig without any other signers (must be at least 1).");
+  CHECK_AND_ASSERT_THROW_MES(threshold > 0 && threshold <= signers.size(),
+    "Tried to make multisig with invalid threshold.");
 
-      // Calculating set of our secret multisig keys as follows: mi = H(Mi),
-      // where mi - secret multisig key, Mi - others' participants public multisig key
-      multisig_keys = cryptonote::calculate_multisig_keys(derivations);
+  // process kex round
+  cryptonote::multisig_account multisig_account;
 
-      // calculating current participant's spend secret key as sum of all secret multisig keys for current participant.
-      // IMPORTANT: participant's secret spend key is not an entire wallet's secret spend!
-      //            Entire wallet's secret spend is sum of all unique secret multisig keys
-      //            among all of participants and is not held by anyone!
-      spend_skey = rct::sk2rct(cryptonote::calculate_multisig_signer_key(multisig_keys));
+  multisig_account.m_threshold = threshold;
+  multisig_account.m_signers = signers;
+  multisig_account.m_aggregate_pubkey = rct::rct2pk(rct::identity());
```

Comment:
Should something like this be in the constructor? In fact several of these values could use a mandatory default.

Reply:
Yes, if `multisig_account` were a dedicated and mature class. Making it a dedicated and mature class should be the subject of a future PR.

Reply:
Why this giant re-write without a few changes necessary to make it mature class. Its completely trivial compared to everything else I had to review for this.

Reply:
...or at the least brace initialize this class when used.